File: sdk-core/src/test/java/care/data4life/sdk/auth/UserServiceTest.java
/*
 * Copyright (c) 2020 D4L data4life gGmbH / All rights reserved.
 *
 * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
 * including any intellectual property rights that subsist in the SDK.
 *
 * The SDK and its documentation may be accessed and used for viewing/review purposes only.
 * Any usage of the SDK for other purposes, including usage for the development of
 * applications/third-party applications shall require the conclusion of a license agreement
 * between you and D4L.
 *
 * If you are interested in licensing the SDK for your own applications/third-party
 * applications and/or if you’d like to contribute to the development of the SDK, please
 * contact D4L by email to help@data4life.care.
 */

package care.data4life.sdk.auth;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;

import care.data4life.auth.AuthorizationService;
import care.data4life.crypto.GCKey;
import care.data4life.crypto.GCKeyPair;
import care.data4life.crypto.GCSymmetricKey;
import care.data4life.sdk.ApiService;
import care.data4life.sdk.CryptoSecureStore;
import care.data4life.sdk.CryptoService;
import care.data4life.sdk.lang.D4LException;
import care.data4life.sdk.network.model.EncryptedKey;
import care.data4life.sdk.network.model.UserInfo;
import care.data4life.sdk.test.util.TestSchedulerRule;
import io.reactivex.Completable;
import io.reactivex.Single;
import io.reactivex.observers.TestObserver;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;

public class UserServiceTest {

    private static final String AUTH_STATE = "authState";
    private static final String USER_ALIAS = "userAlias";
    private static final String KEY_USER_ID = "user_id";
    private static final String AUTH_TOKEN = "authToken";

    @Rule
    public TestSchedulerRule schedulerRule = new TestSchedulerRule();

    private ApiService apiService;
    private CryptoSecureStore storage;
    private UserService userService;
    private CryptoService cryptoService;
    private AuthorizationService authService;

    @Before
    public void setUp() {
        authService = mock(AuthorizationService.class);
        userService = mock(UserService.class);
        storage = mock(CryptoSecureStore.class);
        apiService = mock(ApiService.class);
        cryptoService = mock(CryptoService.class);
        userService = spy(new UserService(USER_ALIAS, authService, apiService, storage, cryptoService));
    }


    @Test
    public void isLoggedIn_shouldReturnTrue() throws Exception {
        // given
        when(authService.isAuthorized(USER_ALIAS)).thenReturn(true);
        doAnswer(invocation -> "auth_state").when(storage).getSecret(AUTH_STATE);
        GCKey mockGCKey = mock(GCKey.class);
        when(mockGCKey.getSymmetricKey()).thenReturn(mock(GCSymmetricKey.class));
        when(cryptoService.fetchTagEncryptionKey()).thenReturn(mockGCKey);
        when(cryptoService.fetchCurrentCommonKey()).thenReturn(mockGCKey);

        // when
        TestObserver<Boolean> testSubscriber = userService.isLoggedIn(USER_ALIAS)
                .test();

        // then
        testSubscriber
                .assertNoErrors()
                .assertValue(true);
    }

    @Test
    public void isLoggedIn_shouldReturnFalse() throws Exception {
        // given
        doAnswer(invocation -> "auth_state").when(storage).getSecret(AUTH_STATE);
        GCKey mockGCKey = mock(GCKey.class);
        when(mockGCKey.getSymmetricKey()).thenReturn(null);
        when(cryptoService.fetchTagEncryptionKey()).thenReturn(mockGCKey);
        when(cryptoService.fetchCurrentCommonKey()).thenReturn(mockGCKey);

        // when
        TestObserver<Boolean> testSubscriber = userService.isLoggedIn(USER_ALIAS)
                .test();

        // then
        testSubscriber
                .assertNoErrors()
                .assertValue(false);
    }

    @Test
    public void getSessionToken_shouldReturnTrue() throws Exception {
        // given
        when(authService.refreshAccessToken(USER_ALIAS)).thenReturn(AUTH_TOKEN);

        // when
        TestObserver<String> testSubscriber = userService.getSessionToken(USER_ALIAS)
                .test();

        // then
        testSubscriber
                .assertNoErrors()
                .assertValue(AUTH_TOKEN);
    }

    @Test
    public void getSessionToken_shouldThrowError() throws Exception {
        // given
        when(authService.refreshAccessToken(USER_ALIAS)).thenReturn(null);

        // when
        TestObserver<String> testSubscriber = userService.getSessionToken(USER_ALIAS)
                .test()
                .await();

        // then
        testSubscriber
                .assertError(NullPointerException.class)
                .assertNotComplete();
    }

    @Test
    public void logout_shouldReturnSuccess() throws InterruptedException {
        // given
        when(apiService.logout(USER_ALIAS)).thenReturn(Completable.complete());

        // when
        TestObserver<Void> testSubscriber = userService.logout()
                .test()
                .await();

        // then
        testSubscriber
                .assertNoErrors()
                .assertComplete();

        verify(storage).clear();
    }

    @Test
    public void logout_shouldReturnException() throws InterruptedException {
        // given
        when(apiService.logout(USER_ALIAS)).thenReturn(Completable.error(new Exception()));

        // when
        TestObserver<Void> testSubscriber = userService.logout()
                .test()
                .await();

        // then
        testSubscriber
                .assertError(Exception.class)
                .assertNotComplete();

        verifyZeroInteractions(storage);
    }

    @Test
    public void finishLogin_shouldReturnBooleanSuccess() {
        // given
        UserInfo userInfo = mock(UserInfo.class);
        when(userInfo.getCommonKey()).thenReturn(mock(EncryptedKey.class));
        when(userInfo.getCommonKeyId()).thenReturn("mockedCommonKeyId");
        when(userInfo.getTagEncryptionKey()).thenReturn(mock(EncryptedKey.class));
        when(userInfo.getUid()).thenReturn("");

        when(apiService.fetchUserInfo(USER_ALIAS)).thenReturn(Single.just(userInfo));

        when(cryptoService.fetchGCKeyPair()).thenReturn(Single.just(mock(GCKeyPair.class)));
        when(cryptoService.asymDecryptSymetricKey(any(), any())).thenReturn(Single.just(mock(GCKey.class)));
        when(cryptoService.symDecryptSymmetricKey(any(), any())).thenReturn(Single.just(mock(GCKey.class)));

        // when
        TestObserver<Boolean> testObserver = userService.finishLogin(true).test();

        // then
        testObserver
                .assertValue(aBoolean -> aBoolean)
                .assertComplete();
    }

    @Test
    public void getUID_shouldReturnString() throws D4LException {
        // given
        String uid = "mock-uid";
        when(storage.getSecret(USER_ALIAS + "_" + KEY_USER_ID, String.class)).thenReturn(uid);

        // when
        TestObserver<String> testObserver = userService.getUID().test();

        // then
        testObserver
                .assertNoErrors()
                .assertValue(uid);
    }
}

File: sdk-core/src/test/java/care/data4life/sdk/auth/UserServiceTest.kt
/*
 * Copyright (c) 2021 D4L data4life gGmbH / All rights reserved.
 *
 * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
 * including any intellectual property rights that subsist in the SDK.
 *
 * The SDK and its documentation may be accessed and used for viewing/review purposes only.
 * Any usage of the SDK for other purposes, including usage for the development of
 * applications/third-party applications shall require the conclusion of a license agreement
 * between you and D4L.
 *
 * If you are interested in licensing the SDK for your own applications/third-party
 * applications and/or if you’d like to contribute to the development of the SDK, please
 * contact D4L by email to help@data4life.care.
 */

package care.data4life.sdk.auth

import care.data4life.sdk.crypto.CryptoContract
import care.data4life.sdk.crypto.GCKey
import care.data4life.sdk.crypto.GCKeyPair
import care.data4life.sdk.log.Log
import care.data4life.sdk.network.NetworkingContract
import care.data4life.sdk.network.model.EncryptedKey
import care.data4life.sdk.network.model.UserInfo
import care.data4life.sdk.test.util.GenericTestDataProvider.ALIAS
import care.data4life.sdk.test.util.GenericTestDataProvider.COMMON_KEY_ID
import care.data4life.sdk.test.util.GenericTestDataProvider.USER_ID
import io.mockk.Runs
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.mockkObject
import io.mockk.unmockkObject
import io.mockk.verify
import io.reactivex.Completable
import io.reactivex.Single
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertSame
import kotlin.test.assertTrue
import org.junit.Before
import org.junit.Test

class UserServiceTest {
    private lateinit var service: AuthContract.UserService
    private val authService: AuthorizationContract.Service = mockk()
    private val apiService: NetworkingContract.Service = mockk()
    private val secureStore: CryptoContract.SecureStore = mockk()
    private val cryptoService: CryptoContract.Service = mockk()

    @Before
    fun setUp() {
        service = UserService(
            ALIAS,
            authService,
            apiService,
            secureStore,
            cryptoService
        )
    }

    @Test
    fun `it fulfils UserService`() {
        val service: Any = UserService(ALIAS, mockk(), mockk(), mockk(), mockk())

        assertTrue(service is AuthContract.UserService)
    }

    @Test
    fun `it resolves the current userId if userId is referenced`() {
        // Given
        val userId = USER_ID

        every {
            secureStore.getSecret(
                "${ALIAS}_user_id",
                String::class.java
            )
        } returns USER_ID

        // When
        val currentId = service.userID.blockingGet()

        // Then
        assertEquals(
            actual = currentId,
            expected = userId
        )
    }

    @Test
    fun `Given isLoggedIn is called with an Alias it returns if the user is logged in`() {
        // Given
        val isLoggedIn = true
        val alias = ALIAS

        every { authService.isAuthorized(alias) } returns isLoggedIn

        // When
        val loggedInState = service.isLoggedIn(alias).blockingGet()

        // Then
        assertEquals(
            actual = loggedInState,
            expected = isLoggedIn
        )
    }

    @Test
    fun `Given isLoggedIn is called with an Alias, it returns false if the authService call fails`() {
        // Given
        val alias = ALIAS

        every { authService.isAuthorized(alias) } throws RuntimeException("Does not matter")

        // When
        val loggedInState = service.isLoggedIn(alias).blockingGet()

        // Then
        assertFalse(loggedInState)
    }

    @Test
    fun `Given logout is called, it returns a Completable while clearing the SecureStore`() {
        // Given
        every { apiService.logout(ALIAS) } returns Completable.fromAction { /* noop */ }
        every { secureStore.clear() } just Runs

        // When
        val loggedInState = service.logout().blockingGet()

        // Then
        assertNull(loggedInState)
        verify(exactly = 1) { secureStore.clear() }
    }

    @Test
    fun `Given logout is called, it returns and logs errors while not clearing the SecureStore if the logout fails`() {
        // Given
        mockkObject(Log)
        val error = RuntimeException("Does not matter")

        every { apiService.logout(ALIAS) } returns Completable.fromAction { throw error }
        every { secureStore.clear() } just Runs
        every { Log.error(error, "Failed to logout") } just Runs

        // When
        val loggedInState = service.logout().blockingGet()

        // Then
        assertSame(
            actual = loggedInState,
            expected = error
        )
        verify(exactly = 0) { secureStore.clear() }
        verify(exactly = 1) { Log.error(error, "Failed to logout") }

        unmockkObject(Log)
    }

    @Test
    fun `Given refreshSessionToken with an Alias, it delegates to call to the AuthService and returns its result`() {
        // Given
        val alias = ALIAS
        val expectedToken = "token"

        every { authService.refreshAccessToken(alias) } returns expectedToken

        // When
        val token = service.refreshSessionToken(alias).blockingGet()

        // Then
        assertEquals(
            actual = token,
            expected = expectedToken
        )
    }

    @Test
    fun `Given finishLogin is called with a Boolean, it returns true`() {
        // Given
        val userInfo: UserInfo = mockk()
        val encryptedCommonKey: EncryptedKey = mockk()
        val encryptedTagEncryptionKey: EncryptedKey = mockk()
        val commonKeyId = COMMON_KEY_ID
        val userId = USER_ID

        val keyPair: GCKeyPair = mockk()
        val commonKey: GCKey = mockk()
        val tagEncryptionKey: GCKey = mockk()

        every { userInfo.userId } returns userId
        every { userInfo.commonKeyId } returns commonKeyId
        every { userInfo.encryptedCommonKey } returns encryptedCommonKey
        every { userInfo.encryptedTagEncryptionKey } returns encryptedTagEncryptionKey

        every { apiService.fetchUserInfo(ALIAS) } returns Single.just(userInfo)
        every { cryptoService.fetchGCKeyPair() } returns Single.just(keyPair)

        every { secureStore.storeSecret("${ALIAS}_user_id", userId) } just Runs

        every {
            cryptoService.asymDecryptSymetricKey(
                keyPair,
                encryptedCommonKey
            )
        } returns Single.just(commonKey)
        every { cryptoService.storeCommonKey(commonKeyId, commonKey) } just Runs
        every { cryptoService.storeCurrentCommonKeyId(commonKeyId) } just Runs

        every {
            cryptoService.symDecryptSymmetricKey(
                commonKey,
                encryptedTagEncryptionKey
            )
        } returns Single.just(tagEncryptionKey)
        every { cryptoService.storeTagEncryptionKey(tagEncryptionKey) } just Runs

        // When
        val result = service.finishLogin(true).blockingGet()

        // Then
        assertTrue(result)

        verify(exactly = 1) { apiService.fetchUserInfo(ALIAS) }
        verify(exactly = 1) { cryptoService.fetchGCKeyPair() }

        verify(exactly = 1) { secureStore.storeSecret("${ALIAS}_user_id", userId) }

        verify(exactly = 1) {
            cryptoService.asymDecryptSymetricKey(
                keyPair,
                encryptedCommonKey
            )
        }
        verify(exactly = 1) {
            cryptoService.storeCommonKey(commonKeyId, commonKey)
        }
        verify(exactly = 1) {
            cryptoService.storeCurrentCommonKeyId(commonKeyId)
        }

        verify(exactly = 1) {
            cryptoService.symDecryptSymmetricKey(
                commonKey,
                encryptedTagEncryptionKey
            )
        }
        verify(exactly = 1) { cryptoService.storeTagEncryptionKey(tagEncryptionKey) }
    }

    @Test
    fun `Given finishLogin is called with a Boolean and any operation fails, it returns true while logging the Error`() {
        // Given
        mockkObject(Log)

        val error = RuntimeException("not important")

        val userInfo: UserInfo = mockk()
        val encryptedCommonKey: EncryptedKey = mockk()
        val encryptedTagEncryptionKey: EncryptedKey = mockk()
        val commonKeyId = COMMON_KEY_ID
        val userId = USER_ID

        val keyPair: GCKeyPair = mockk()
        val commonKey: GCKey = mockk()
        val tagEncryptionKey: GCKey = mockk()

        every { userInfo.userId } returns userId
        every { userInfo.commonKeyId } returns commonKeyId
        every { userInfo.encryptedCommonKey } returns encryptedCommonKey
        every { userInfo.encryptedTagEncryptionKey } returns encryptedTagEncryptionKey

        every { apiService.fetchUserInfo(ALIAS) } returns Single.just(userInfo)
        every { cryptoService.fetchGCKeyPair() } returns Single.just(keyPair)

        every { secureStore.storeSecret("${ALIAS}_user_id", userId) } just Runs

        every {
            cryptoService.asymDecryptSymetricKey(
                keyPair,
                encryptedCommonKey
            )
        } returns Single.error(error)
        every { cryptoService.storeCommonKey(commonKeyId, commonKey) } just Runs
        every { cryptoService.storeCurrentCommonKeyId(commonKeyId) } just Runs

        every {
            cryptoService.symDecryptSymmetricKey(
                commonKey,
                encryptedTagEncryptionKey
            )
        } returns Single.just(tagEncryptionKey)
        every { cryptoService.storeTagEncryptionKey(tagEncryptionKey) } just Runs

        every { Log.error(error, "Failed to finish login") } just Runs

        // Then
        val result = assertFailsWith<RuntimeException> {
            // When
            service.finishLogin(true).blockingGet()
        }

        assertSame(
            actual = result,
            expected = error
        )

        verify(exactly = 1) { apiService.fetchUserInfo(ALIAS) }
        verify(exactly = 1) { cryptoService.fetchGCKeyPair() }

        verify(exactly = 1) { secureStore.storeSecret("${ALIAS}_user_id", userId) }

        verify(exactly = 1) {
            cryptoService.asymDecryptSymetricKey(
                keyPair,
                encryptedCommonKey
            )
        }
        verify(exactly = 0) {
            cryptoService.storeCommonKey(commonKeyId, commonKey)
        }
        verify(exactly = 0) {
            cryptoService.storeCurrentCommonKeyId(commonKeyId)
        }

        verify(exactly = 0) {
            cryptoService.symDecryptSymmetricKey(
                commonKey,
                encryptedTagEncryptionKey
            )
        }
        verify(exactly = 0) { cryptoService.storeTagEncryptionKey(tagEncryptionKey) }
        verify(exactly = 1) { Log.error(error, "Failed to finish login") }

        mockkObject(Log)
    }
}
